src/app.ts

```diff
@@ -65,62 +34,36 @@ const dbusInvoker = (message: Message): PromiseLike<any> => {
 	});
 };
 
-/**
- * Retrieves the IPv4 address for the named interface.
- *
- * @param intf Name of interface to query
- */
-const getNamedInterfaceAddr = (intf: string): string => {
-	const nics = os.networkInterfaces()[intf];
-
-	if (!nics) {
-		throw new Error('The configured interface is not present, exiting');
-	}
-
-	// We need to look for the IPv4 address
-	let ipv4Intf;
-	for (const nic of nics) {
-		if (nic.family === 'IPv4') {
-			ipv4Intf = nic;
-			break;
-		}
-	}
-
-	if (!ipv4Intf) {
-		throw new Error(
-			'IPv4 version of configured interface is not present, exiting',
-		);
+const getAddrs4 = (
+	iface?: string,
+): Array<os.NetworkInterfaceInfo | undefined> => {
+	if (iface) {
+		return Object.values([os.networkInterfaces()[iface]]).flatMap((infos) => {
+			return infos?.find((i) => i.family === 'IPv4' && !i.internal);
+		});
+	} else {
+		return Object.values(os.networkInterfaces()).flatMap((infos) => {
+			return infos?.find((i) => i.family === 'IPv4' && !i.internal);
+		});
 	}
```

Comment:
If you put everything in one chain, the undefined in the type should go away
```suggestion
	Object.entries(os.networkInterfaces())
	  .filter(([nic,ips]) => !iface || nic === iface)
	  .flatMap(([,ips])=>ips)
	  .filter(ip=>!ip.internal && ip.family === "IPv4")
```

Reply:
similar change needed below of course

Reply:
@Hades32 so something like:

```
const getAddrs6 = (
	iface?: string,
): Array<os.NetworkInterfaceInfo | undefined> => {
	return Object.entries(os.networkInterfaces())
	  .filter(([nic,ips]) => !iface || nic === iface)
	  .flatMap(([,ips])=>ips)
	  .filter(ips=>!ip.internal && ip.family === "IPv6" && ip.scopeid === 0)
};
```

Reply:
yes, but we don't need the ` | undefined` anymore

Reply:
@Hades32 it's angry...

```
> @1.9.14 build /Users/anton/dev/balena-io/balena-mdns-publisher
> tsc
                                                                    
src/app.ts:40:2 - error TS2322: Type '(NetworkInterfaceInfoIPv4 | NetworkInterfaceInfoIPv6 | undefined)[]' is not assignable to type 'Ne
tworkInterfaceInfo[]'.                                                                                                                  
  Type 'NetworkInterfaceInfoIPv4 | NetworkInterfaceInfoIPv6 | undefined' is not assignable to type 'NetworkInterfaceInfo'.
    Type 'undefined' is not assignable to type 'NetworkInterfaceInfo'.
                                                                    
40  return Object.entries(os.networkInterfaces())
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                       
41    .filter(([nic,ips]) => !iface || nic === iface)
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
42    .flatMap(([,ips])=>ips)                                       
   ~~~~~~~~~~~~~~~~~~~~~~~~~~
43    .filter(ip=>!ip.internal && ip.family === "IPv4")                                                                                 
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~     

src/app.ts:41:18 - error TS6133: 'ips' is declared but its value is never read.                   
                                                                                                                                        
41    .filter(([nic,ips]) => !iface || nic === iface)
                    ~~~                                                                                                                 
                                                                                                                                        
src/app.ts:43:17 - error TS2532: Object is possibly 'undefined'.

43    .filter(ip=>!ip.internal && ip.family === "IPv4")
                   ~~
                                  
src/app.ts:43:32 - error TS2532: Object is possibly 'undefined'.
                                  
43    .filter(ip=>!ip.internal && ip.family === "IPv4")
                                  ~~
...
```

Reply:
@ab77 seems it's valid to have IPs at all. We can just replace that case with an empty array:
```typescript
let ips = Object.entries(os.networkInterfaces())
      .filter(([nic]) => (!iface || nic === iface))
      .flatMap(([, ips]) => ips || [])
      .filter((ip) => !ip.internal && ip.family === 'IPv4');
```

Reply:
@Hades32 that got rid of most of the red, only have this now:

```
src/app.ts:39:18 - error TS6133: 'ips' is declared but its value is never read.

39      .filter(([nic, ips]) => !iface || nic === iface)
                       ~~~
```

I tried `noUnusedLocals: true|false` in tsconfig, doesn't make any difference...

Reply:
You should be able to just skip the argument or mark it as ignored by prefixing it with an underscore. Don't mess with tsconfig -- it's (usually) set up sensibly.

Option A (preferred):

```
.filter(([nic]) => !iface || nic === iface)
```

Option B:

```
.filter(([nic, _ips]) => !iface || nic === iface)
```

Reply:
@dfunckt that was easy, thanks.